### PR TITLE
fetcher: add workflow specification fetcher

### LINKED ISSUE #441: implement flexible workflow specification fetcher
When running workflows defined in external sources, there is a need to fetch workflow specification. See parent issue https://github.com/reanahub/reana-server/issues/440

This issue concerns designing and implementing a flexible workflow specification fetcher for a variety of sources:

- if git, then do shallow clone such as `git clone --depth=1` and use `reana.yaml` discovered there
- if URL to exact `reana.yaml` location, then `wget` it
- if Zenodo tarball, then fetch tarball and "discover" `reana.yaml` hidden inside it

We could come up with an abstract `WorkflowSpecificationFetcher` class and subclasses implementing various fetching techniques.

Note that for Serial workflows it may be enough to just deal with `reana.yaml`. But for CWL/Snakemake/Yadage we may need to fetch also `workflow.yaml`, `steps.yaml`, and `inputs.yaml` and friends. (Easy by `git clone`; harder when having only URL since we would have to discover where `steps.yaml` etc are by reading `reana.yaml`.)

The expected outcome of the fetcher is having the files necessary to run the workflow present in a desired output directory. (Which will be equal to workspace location.)

Note: we are not concerned here with how we store the workflow specs in the REANA database etc. The focus is on implementing a flexible component that can run even outside of REANA. We shall plug it into REANA flow as par of the parent issue.

### PR DESCRIPTION
**Pull request by mdonadoni**
Closes #441

How to test:
1. Save the following python snippet to a file in the repository's root directory (e.g. `pr445.py`)
```python
import os
import random
import string

from reana_server.fetcher import get_fetcher
from tests.test_fetcher import GIT_URL, GITHUB_REPO_URL, YAML_URL, ZENODO_URL


def random_string():
    return "".join(random.choices(string.ascii_lowercase, k=5))


def fetch(url, tmp_dir=None):
    if not tmp_dir:
        tmp_dir = os.path.join("/tmp", "test-fetcher-{}".format(random_string()))
        os.makedirs(tmp_dir)
    print(f"Fetching to directory: {tmp_dir}")
    fetcher = get_fetcher(url, tmp_dir)
    fetcher.fetch()
    spec = fetcher.workflow_spec_path()
    print(f"REANA workflow specification file: {spec}")

```
2. Execute the saved file in interactive mode: `python -i pr445.py`
3. Call the `fetch` function passing an URL and an output directory. If the output directory is not provided, a random one in `/tmp` will be used

Expected results: The workflow specification is fetched correctly. When fetching from a Git repository or a zip archive, also the other files will be present in the output directory.
```pycon
>>> fetch(GIT_URL)
Fetching to directory: /tmp/test-fetcher-fyjka
REANA workflow specification file: /tmp/test-fetcher-fyjka/reana.yaml
>>> fetch(GITHUB_REPO_URL)
Fetching to directory: /tmp/test-fetcher-ohhav
REANA workflow specification file: /tmp/test-fetcher-ohhav/reana.yaml
>>> fetch(YAML_URL)
Fetching to directory: /tmp/test-fetcher-satvh
REANA workflow specification file: /tmp/test-fetcher-satvh/reana.yaml
>>> fetch(ZENODO_URL)
Fetching to directory: /tmp/test-fetcher-lsxdp
REANA workflow specification file: /tmp/test-fetcher-lsxdp/reana.yaml
```

### CODE REVIEW

File: reana_server/fetcher.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    # 3. URL to a specific folder: /<user>/<repo>/tree/<git_ref>/<path_to_dir>
+    # 4. URL to a specific file: /<user>/<repo>/blob/<git_ref>/<path_to_file>
+
+    # We are interested in five components: username, repository, tree/blob, git_ref, path
+    NUM_COMPONENTS = 5
+    parsed_url = urlparse(url)
+    url_path_components = parsed_url.path.strip("/").split(
+        "/", maxsplit=NUM_COMPONENTS - 1
+    )
+    url_path_components = url_path_components + [None] * (
```
[the remark this answers is not in the document]
Comment (mvidalgarcia) [reply reana_server/fetcher.py:197]: OK, I understand now... IMHO I find it quite hacky and not very easy to read.. It might be worth trying to find another way of solving this or at least leave a comment.

[the remark this answers is not in the document]
Comment (mvidalgarcia) [reply reana_server/fetcher.py:197]: Not much better.. but I came up with this solution:

```python
def get_url_path_components():
  url_bits = parsed_url.path.strip("/").split("/", maxsplit=NUM_COMPONENTS - 1)
  for i in range(NUM_COMPONENTS):
      try:
          yield url_bits[i]
      except IndexError:
          yield None
```

Not very happy about it though, you can ignore if you don't like it.

File: tests/test_fetcher.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        (GITHUB_REPO_URL, "reana.yaml"),
+        (GITHUB_REPO_URL + "/blob/master/reana-snakemake.yaml", "reana-snakemake.yaml"),
+        (
+            GITHUB_REPO_URL + "/tree/1fbe52138b98a85a15fde58d385b21b637c18be8",
+            "reana.yaml",
+        ),
+        pytest.param(
+            ZENODO_URL,
+            "reana.yaml",
+            marks=pytest.mark.skip(reason="big file to download"),
```
[the remark this answers is not in the document]
Comment (mdonadoni) [reply tests/test_fetcher.py:54]: This test is being skipped because it downloads a quite big zip file (~50MB), which might take a while if the connection is not very fast. Do you think we should run this test each time anyway?

[the remark this answers is not in the document]
Comment (mvidalgarcia) [reply tests/test_fetcher.py:54]: Mmmh I'm not even sure if we should rely on downloading stuff from those URLs. Even if they should be rather stable, they might change at some point, which would break all the tests. In other words, I'm not sure that unit tests should depend on requesting network resources ie. having an Internet connection at all. Perhaps @tiborsimko has a better-formed opinion about this.

My question was more about, if you skip the test (big zip file is a valid reason, of course), then what's the relevance of keeping this test? Is there any way to force its execution? In other words, under which circumstances that particular test will be executed?

[the remark this answers is not in the document]
Comment (mdonadoni) [reply tests/test_fetcher.py:54]: I decided to keep this test instead of just omitting it because I think it makes it clear that we are skipping the test only because it is time/bandwidth-consuming and not because of some other reason. In any case, it shouldn't be difficult to rewrite these tests to avoid network calls.

**Comment by mdonadoni** [comment]
I have updated the PR description with some steps on how to test this new functionality

File: tests/test_fetcher.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tiborsimko) [reply tests/test_fetcher.py:60]: What about putting those test zip files etc into `pytest-reana` package and use them here as fixtures instead of downloading them from external URLs?  We have obviously local access to `pytest-reana` when we run tests, so we wouldn't be dependent on any external service, so avoid problems you mentioned above such as (i) network resource troubles but also (ii) what-if-author-changes-zenodo-record troubles.

If our test zip file is small, i.e. we delete most of the regular source file content and keep only `reana.yaml` there, then it will nicely fit into PyPI package as `include_package_data`, so it could be practical solution. And we could mock URL loading from Zenodo but fetch the file from there.

TL;DR WDYT about treating examples as other pytest fixtures?

[the remark this answers is not in the document]
Comment (mvidalgarcia) [reply tests/test_fetcher.py:60]: I think it's not only zip files, but all that's hosted  behind these links:

```
GIT_URL = "https://github.com/reanahub/reana-demo-root6-roofit.git"
GITHUB_REPO_URL = "https://github.com/reanahub/reana-demo-root6-roofit"
ZENODO_URL = "https://zenodo.org/record/5752285/files/circular-health-data-processing-master.zip?download=1"
YAML_URL = "https://raw.githubusercontent.com/reanahub/reana-demo-root6-roofit/master/reana.yaml"
```

If I'm not mistaken, all this content is downloaded to perform the tests.

**Review by mvidalgarcia — APPROVED** [review]
LGTM, good work 👏
